### PR TITLE
Use ResXFileRef from this project

### LINKED ISSUE #5: Can't load XML resource
When an XML file is added to Resources in Visual Studio, the following XML is stored in the resx file:

```xml
  <assembly alias="System.Windows.Forms" name="System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
  <data name="Test" type="System.Resources.ResXFileRef, System.Windows.Forms">
    <value>Test.xml;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
  </data>
```

Trying to load this data (in a `net7.0` application) throws the following exception:

```
System.ArgumentException: ResX file Could not load file or assembly 'System.Windows.Forms, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified. Line 8015, position 5. cannot be parsed.
 ---> System.Xml.XmlException: Could not load file or assembly 'System.Windows.Forms, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified. Line 8015, position 5.
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Windows.Forms, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
File name: 'System.Windows.Forms, Culture=neutral, PublicKeyToken=null'
   at System.Reflection.RuntimeAssembly.InternalLoad(AssemblyName assemblyName, StackCrawlMark& stackMark, AssemblyLoadContext assemblyLoadContext, RuntimeAssembly requestingAssembly, Boolean throwOnFileNotFound)
   at System.Reflection.Assembly.Load(String assemblyString)
   at System.Resources.NetStandard.AssemblyNamesTypeResolutionService.GetAssembly(AssemblyName name, Boolean throwOnError)
   at System.Resources.NetStandard.AssemblyNamesTypeResolutionService.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
   at System.Resources.NetStandard.ResXDataNode.InitializeDataNode(String basePath)
   at System.Resources.NetStandard.ResXDataNode..ctor(DataNodeInfo nodeInfo, String basePath)
   at System.Resources.NetStandard.ResXResourceReader.ParseDataNode(XmlTextReader reader, Boolean isMetaData)
   at System.Resources.NetStandard.ResXResourceReader.ParseXml(XmlTextReader reader)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Resources.NetStandard.ResXResourceReader.ParseXml(XmlTextReader reader)
   at System.Resources.NetStandard.ResXResourceReader.EnsureResData()
   at System.Resources.NetStandard.ResXResourceReader.GetEnumerator()
```

It seems like this could work if ResXResourceReader.NetStandard used its own implementation of `ResXFileRef` instead of trying to load the WinForms one.

### PR DESCRIPTION
**Pull request by bgrainger**
Fixes #5

Detect the fully-qualified type name of the WinForms `ResXFileRef` type and substitute the one from this project instead.

### CODE REVIEW

File: System.Resources.NetStandard/ResXDataNode.cs [hunk 1]
```diff
@@ -956,6 +956,13 @@ public Type GetType(string name, bool throwOnError, bool ignoreCase)
                 return result;
             }
 
+            // hard-code the assembly name for ResXFileRef and replace it with our own type instead of loading the one from System.Windows.Forms
+            if (name == "System.Resources.ResXFileRef, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")
```
Comment (farlee2121) [inline System.Resources.NetStandard/ResXDataNode.cs]: Could you move this string to [`ResXConstants`](https://github.com/farlee2121/ResXResourceReader.NetStandard/blob/main/System.Resources.NetStandard/ResxConstants.cs)? 
It's nice to keep all the magic strings in one place.

Also, I noticed the example Resx I made with the VS editor didn't have this whole string as it's type
```xml
<data name="Test" type="System.Resources.ResXFileRef, System.Windows.Forms">
  <value>TestResources\Files\Test.xml;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
</data>
```
Perhaps we can match on a StartsWith `System.Resources.ResXFileRef, System.Windows.Forms` (but still keep the constant in `ResXConstants`

File: Tests/ResxDataNodeTests.cs [hunk 2]
```diff
@@ -18,5 +20,31 @@ public void ResxDataNode_ResXFileRefConstructor()
             Assert.Equal(nodeName, dataNode.Name);
             Assert.Same(fileRef, dataNode.FileRef);
         }
+
+        [Fact]
+        public void ResxDataNode_CreateForResXFileRef()
+        {
+            // Simulate an XML file stored in resources, which uses a ResXFileRef. The actual resx XML looks like:
```
Comment (farlee2121) [inline Tests/ResxDataNodeTests.cs:35]: I appreciate the explanation

File: Tests/ResxDataNodeTests.cs [hunk 3]
```diff
@@ -18,5 +20,31 @@ public void ResxDataNode_ResXFileRefConstructor()
...
+               <assembly alias="System.Windows.Forms" name="System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+               <data name="Test" type="System.Resources.ResXFileRef, System.Windows.Forms">
+                    <value>Test.xml;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
+                </data>
+            */
+            var nodeInfo = new DataNodeInfo
+            {
+                Name = "Test",
+                ReaderPosition = new Point(1, 2),
+                TypeName = "System.Resources.ResXFileRef, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
```
Comment (farlee2121) [inline Tests/ResxDataNodeTests.cs]: This TypeName could also lean on the ResXConstants value once moved. Consolidates the special strings a bit.

File: System.Resources.NetStandard/ResXDataNode.cs (hunk 1 above)
Comment (farlee2121) [reply System.Resources.NetStandard/ResXDataNode.cs]: Hmm. This also causes the namespace to change when we write the resx.
1. We should probably make sure we can still read the data after we write the changed namespace
2.  This breaks compatibility with visual editors. Not sure what to do about that

**Comment by farlee2121** [comment]
Overall, I see the issue and I think I understand the solution.
My main concern is the broken compatibility with Visual Studio tooling.

I'm not quite sure how else we solve it though, seeing as we can't reference winforms from net standard...

**Comment by bgrainger** [comment]
> My main concern is the broken compatibility with Visual Studio tooling.

Can you explain what you mean here? I wanted this change to enable this library to be able to read a resx file (that was generated by Visual Studio) without requiring all the WinForms dependencies. (That is, to load the resx XML without requiring any change in the type name that was serialised in the XML.) In my testing, enumerating resources from a .NET Framework app produced the exact same key/value pairs as this library with the PR, so it seemed 100% compatible.

I notice that the tests are broken on non-Windows; I'm not sure why that is but can take a look.

**Comment by farlee2121** [comment]
Your change fixes the library's ability to read Resx files with file references.
But if we then write back to that file, visual studio can't edit that file with it's visual Resx editor anymore.

working vs resx editor

broken vs resx editor

The second image is using the file produced from this file roundtrip test. I adapted this from your test to make sure we could read files we write, considering the ResXFileRef namespace is changed when we resolve types.
```cs
[Fact]
public void ResxDataNode_ResXFileRef_RoundTrip()
{
    var referencedFileContent = Example.FileRef;
    
    var nodeInfo = new DataNodeInfo
    {
        Name = "Test",
        ReaderPosition = new Point(1, 2),
        TypeName = "System.Resources.ResXFileRef, System.Windows.Forms",
        ValueData = "TestResources/Files/FileRef.xml;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8",
    };
    var dataNode = new ResXDataNode(nodeInfo, null);
    var typeResolver = new AssemblyNamesTypeResolutionService(Array.Empty<AssemblyName>());
    
    string resxPath = @".\nya-test.resx";
    using (ResXResourceWriter writer = new ResXResourceWriter(resxPath))
    {
        writer.AddResource(dataNode);
    }
    using(ResXResourceReader reader = new ResXResourceReader(resxPath))
    {
        var dictionary = new Dictionary<object, object>();
        IDictionaryEnumerator dictionaryEnumerator = reader.GetEnumerator();
        while (dictionaryEnumerator.MoveNext())
        {
            dictionary.Add(dictionaryEnumerator.Key, dictionaryEnumerator.Value);
        }

        Assert.Equal(referencedFileContent, dictionary.GetValueOrDefault(nodeInfo.Name));
    }
}
```

**Comment by farlee2121** [comment]
`Example.FileRef` is essentially just a Resx getter for the Test.xml file you provided, that way we can compare the file contents without copying them to code

**Comment by farlee2121** [comment]
@bgrainger Are you still interested in collaborating on this?

**Comment by bgrainger** [comment]
Yes, I am. Are you waiting for me to solve the "saving back to XML" problem?

**Comment by farlee2121** [comment]
I suppose I was waiting for some kind of feedback. 
Solving the "saving back to XML" problem is something I think we should do before releasing.
I happy to collaborate, I just don't want to take over your contribution.

**Comment by bgrainger** [comment]
Sorry for the radio silence.

I'm still interested in having this problem be solved. If you have a solution in mind and know the way forward to implement it, I'm happy for you to take it from here (and use as much or as little of my code as makes sense).

If you don't have the bandwidth to take it on, I'm happy to look at it more (maybe after American Thanksgiving).

**Comment by farlee2121** [comment]
Sounds good. I pull requested code for some my comments back to your branch.

**Comment by farlee2121** [comment]
Whenever you're back from vacation, I think I found a good solution to the write issue. I added it to the PR to your branch.

However, I also noticed that we'll run into issues if anyone overwrites the `ITypeResolutionService`. I'm not sure how likely that is...

I could use some help kicking the tires and poking holes in the test suite.

**Comment by bgrainger** [comment]
I incorporated your changes and added a couple of fixes that I needed to get the tests working locally.

**Comment by farlee2121** [comment]
Weird, I thought I had replaced the test file with a StringWriter

File: Directory.Build.props [hunk 4]
```diff
@@ -14,5 +14,6 @@
 	  <PackageLicenseExpression>MIT</PackageLicenseExpression>
 	  <RepositoryUrl>https://github.com/farlee2121/ResXResourceReader.NetStandard</RepositoryUrl>
 	  <PackageTags>ResX;ResXResourceReader;System.Resources;ResXResourceWriter;netstandard</PackageTags>
+	  <LangVersion>9.0</LangVersion>
```
Comment (farlee2121) [inline Directory.Build.props]: This appears to be breaking the build

```
CSC : error CS1617: Invalid option '9.0' for /langversion. Use '/langversion:?' to list supported values. [D:\a\ResXResourceReader.NetStandard\ResXResourceReader.NetStandard\System.Resources.NetStandard\System.Resources.NetStandard.csproj]
CSC : error CS1617: Invalid option '9.0' for /langversion. Use '/langversion:?' to list supported values. [D:\a\ResXResourceReader.NetStandard\ResXResourceReader.NetStandard\System.Resources.NetStandard.StrongName\System.Resources.NetStandard.StrongName.csproj]
CSC : error CS1617: Invalid option '9.0' for /langversion. Use '/langversion:?' to list supported values. [D:\a\ResXResourceReader.NetStandard\ResXResourceReader.NetStandard\System.Resources.NetStandard.StrongName\System.Resources.NetStandard.StrongName.csproj]
CSC : error CS1617: Invalid option '9.0' for /langversion. Use '/langversion:?' to list supported values. [D:\a\ResXResourceReader.NetStandard\ResXResourceReader.NetStandard\System.Resources.NetStandard\System.Resources.NetStandard.csproj]
```

File: Tests/ResxDataNodeTests.cs [hunk 5]
```diff
@@ -18,5 +28,135 @@ public void ResxDataNode_ResXFileRefConstructor()
...
+                var dataNodes = ReaderToNodes(reader);
+
+                using (ResXResourceWriter writer = new ResXResourceWriter(new StringWriter(writerOutput)))
+                {
+                    dataNodes.ForEach(writer.AddResource);
+                    writer.Generate();
+                }
+            }
+
+            Assert.Equal(originalResx, writerOutput.ToString().Replace("\r\n", "\n"));
```
Comment (farlee2121) [inline Tests/ResxDataNodeTests.cs]: This breaks the test on my computer. We can't really write it to favor a particular platform line ending. 

Maybe you could use something like [ReplaceLineEndings](https://stackoverflow.com/a/75133341/2436737). 
We should be able to update the test project's target framework version if need be.

File: Directory.Build.props (hunk 4 above)
Comment (farlee2121) [reply Directory.Build.props]: Since we're bound to netstandard, we probably won't be able to leverage newer language features.

**Review by farlee2121 — APPROVED** [review]

File: Directory.Build.props (hunk 4 above)
Comment (bgrainger) [reply Directory.Build.props]: Most newer C# features can be used with netstandard or .NET Framework. However, they do require a newer .NET SDK to build (which isn't installed by the current CI script). I reverted this change.

**Comment by farlee2121** [comment]
@bgrainger I just pushed out new package versions

**Pull request merged**